### PR TITLE
Fixes #16449: fuzzy matching to help with error messages

### PR DESCRIPTION
**Pull request by gpoblon**
https://issues.rudder.io/issues/16449

Compiling and implemented fuzzy string matching feature which is a useful tool to have when dealing with compilation errors such as type or variable not found. Finds a similar variable/type name available in the actual context (global + current scope) and indicates similar names.

### CODE REVIEW

File: rudder-lang/src/error.rs [hunk 1]
```diff
@@ -205,3 +209,50 @@ impl fmt::Display for Error {
         }
     }
 }
+
+
+// Must have a trait to add the following method as ngrammatic::corpus only has an `add_text()` method
+trait FuzzyChecker<'src, I> {
+	fn add_vec(&mut self, list: I);
```
Comment (peckpeck) [inline rudder-lang/src/error.rs]: you call it add_vec but you don't add a vector ?

Comment (peckpeck) [reply rudder-lang/src/error.rs]: You have a generic parameter here but you only have a single implementation for an iterator

File: rudder-lang/src/error.rs [hunk 2]
```diff
@@ -205,3 +209,50 @@ impl fmt::Display for Error {
...
+}
+
+impl<'src, I: Iterator<Item = &'src Token<'src>>> FuzzyChecker<'src, I> for Corpus {
```
Comment (peckpeck) [inline rudder-lang/src/error.rs]: It works here, but usually having a life time "&'a Object<'a>" often ends up with the constraint beging 'a == 'static 
which can become problematic, so avoid it when possible

File: rudder-lang/src/error.rs [hunk 3]
```diff
@@ -205,3 +209,50 @@ impl fmt::Display for Error {
...
+}
+
+/// Searches for a matching string in an Iterator of Token
+fn fuzzy_search<'src, I>(token_fragment: &str, list: I) -> Option<String>
+where 
+	I: Iterator<Item = &'src Token<'src>>,
+{
+	let mut corpus = CorpusBuilder::new().finish();
+	corpus.add_vec(list);
+    let results = corpus.search(token_fragment, FUZZY_THRESHOLD);
```
Comment (peckpeck) [inline rudder-lang/src/error.rs]: indent

File: rudder-lang/src/ast/enums.rs [hunk 4]
```diff
@@ -195,10 +195,11 @@ impl<'src> EnumList<'src> {
                     if !(values.contains(&f) || *f == "*") {
                         fail!(
                             f,
-                            "Value {} used in mapping {} does not exist in {}",
+                            "Value {} used in mapping {} does not exist in {}{}",
```
Comment (peckpeck) [inline rudder-lang/src/ast/enums.rs]: Missing separator

File: rudder-lang/src/ast/enums.rs [hunk 5]
```diff
@@ -379,11 +386,17 @@ impl<'src> EnumList<'src> {
                         } else {
                             match self.enums.get(&e1) {
                                 // or get it from a global enum
-                                None => fail!(e1, "No such enum {}", e1),
+                                None => fail!(
+                                    e1,
+                                    "No such enum {}{}",
```
Comment (peckpeck) [inline rudder-lang/src/ast/enums.rs]: Missing separator

File: rudder-lang/src/ast.rs [hunk 6]
```diff
@@ -311,29 +314,68 @@ impl<'src> AST<'src> {
         }
     }
 
+    // binding_check fn dependance: compares library and user function's parameters. if diff, output an error
```
Comment (amousset) [inline rudder-lang/src/ast.rs]: Should probably be a doc comment `///`

**Pull request merged**